File: ruby/kindergarten-garden/kindergarten_garden.rb
class Garden
  def initialize(diagram, students = STUDENTS)
    rows(diagram).zip(students.sort) do |(top, bottom), student|
      define_singleton_method student.downcase do
        top.concat(bottom)
      end
    end
  end

  private

  def rows(diagram)
    diagram.split("\n").map do |row|
      row.chars.map { |code| ENCODED_PLANTS[code] }.each_slice(2)
    end.then do |(top, bottom)|
      top.zip(bottom)
    end
  end

  STUDENTS = %w[Alice Bob Charlie David Eve Fred Ginny Harriet Ileana Joseph Kincaid Larry].freeze
  ENCODED_PLANTS = {
    'V' => :violets,
    'C' => :clover,
    'G' => :grass,
    'R' => :radishes
  }.freeze
  private_constant :STUDENTS, :ENCODED_PLANTS
end

File: kotlin/kindergarten-garden/src/main/kotlin/KindergartenGarden.kt
val PLANTS = arrayOf("violets", "clover", "radishes", "grass")
val STUDENTS = arrayOf("Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry")

class KindergartenGarden(diagram: String) {
    val plantsForStudent = diagram.lines()
        .map { line -> line.map { code -> PLANTS.first { it[0].uppercaseChar() == code } } }
        .let { it[0].chunked(2).zip(it[1].chunked(2), { top, bottom -> top + bottom }) }
        .mapIndexed { i, plants -> STUDENTS[i] to plants}
        .toMap()

    fun getPlantsOfStudent(student: String) = plantsForStudent[student].orEmpty()
}
